### PR TITLE
Block Styles: Reduce specificity of block style variation selectors

### LINKED ISSUE #61268: Block Styles: Reducing specificity of default block style variation styles
Part of: https://github.com/WordPress/gutenberg/issues/57537

**Note: This is a blocker for the extended block style variations feature and [section styling](https://github.com/WordPress/gutenberg/issues/57537).**

## Problem 

To support applying block style variations in a nested fashion, the specificity of block style variation selectors needs to be `0-0-0`. 

The Button block's stylesheet includes some default styles for the Outline block style variation to make its background transparent. If the selector for the Outline style's transparent background is reduced to zero it gets overridden by any button element or Button block styles in global styles. 

This is due to the current style load order which is:
1. block library styles including defaults for block style variations,
2. global styles,
3. and finally in [#57908](https://github.com/WordPress/gutenberg/pull/57908) stylesheets for the application of block style variations.

Additional context can be found in https://github.com/WordPress/gutenberg/pull/61032#issuecomment-2087867699.

## Potential Solutions

The ideal order of styles here would be:

1. Block library styles
2. Global Styles
3. Defaults for block style variations e.g. Button Outline's transparent background
4. Block style variation styles

To achieve this, adding support for a new stylesheet for blocks would be beneficial. Perhaps something like `variations.scss` for block library blocks which can then be collected and enqueued in the appropriate order.

### Alternative Option

One long term alternative could be to extend the `styles` property within the block.json API to allow them to define styles for block style variations. This would allow for dynamically generating these default styles for variations and outputting them when and if they make sense.

This dynamic approach however might not be a silver bullet solution given the potential for style variations to need styling that theme.json does not yet support. It would also need extra work to ensure compatibility across classic and block themes.

### PR DESCRIPTION
**Pull request by aaronrobertshaw**
Fixes: https://github.com/WordPress/gutenberg/issues/61268

Related:
- https://github.com/WordPress/gutenberg/pull/57908

_**Note: The changes in this PR are being split out from https://github.com/WordPress/gutenberg/pull/57908 to make that more manageable**_

## What?

Reduces the specificity of block style variation selectors.

This is a continuation of https://github.com/WordPress/gutenberg/pull/57659 and the overall reduction of global styles specificity in https://github.com/WordPress/gutenberg/pull/57841.

## Why?

To nest extended block style variations, as proposed in #57908, we need to reduce the specificity of related selectors to `0-0-0`. This allows the load order of styles to determine the final cascade. 

This enables the ability for a block style variation to define inner block type styles without preventing an inner block from applying its own desired block style variation.

#### What happens if specificity isn't reduced?

Imagine a group block style variation that adds a border to inner image blocks. In this case, we'd end up with a selector such as `.wp-block-group.is-style-bordered-images .wp-block-image img` and a specificity of `0-3-1`.

If we wanted to apply a rounded non-bordered image block style variation to a single inner image block, it would get a selector like `.wp-block-image.is-style-rounded-without-border img` with `0-2-1` specificity. 

The result is the outer variation's `img` styles incorrectly take precedence.

#### Alternatives

Explorations were made into tweaking style variation selectors with `:not` clauses to try and exclude inner blocks that have variations applied. The results here were brittle, incredibly complex, and only really worked for one level of nesting, leading back to the approach proposed in this PR.

## Testing Instructions

1. Check core block styles are still applied correctly in the editor and frontend e.g. button outline, image rounded etc.
2. `npm run test:unit packages/block-editor/src/components/global-styles/test/use-global-styles-output.js`
3. `npm run test:unit:php:base -- --filter WP_Theme_JSON_Gutenberg_Test`

### CODE REVIEW

**Comment by talldan** [comment]
The rounded style for the image block looks a little different in this PR compared to trunk.

This PR:
Screenshot 2024-04-26 at 4 15 22 pm

Trunk:
Screenshot 2024-04-26 at 4 15 01 pm

In this PR, this style rule seems to be taking precedence:
```css
.wp-block-image .is-style-rounded img, .wp-block-image.is-style-circle-mask img, .wp-block-image.is-style-rounded img {
    border-radius: 9999px;
}
```

While in trunk it's this one:
```css
.wp-block-image.is-style-rounded img, .wp-block-image.is-style-rounded .wp-block-image__crop-area, .wp-block-image.is-style-rounded .components-placeholder {
    border-radius: var(--wp--preset--spacing--20);
}
```

I'm not sure why there are two different 'rounded' styles 🤔

**Comment by aaronrobertshaw** [comment]
Thanks for flagging that @talldan 👍 

> I'm not sure why there are two different 'rounded' styles 🤔

TwentyTwentyFour customizes the core image block's rounded style [via theme.json](https://themes.trac.wordpress.org/browser/twentytwentyfour/1.1/theme.json#L506).

I'll need to drop the specificity of the style within the Image block's stylesheet. I mistakenly thought this had already been done in other recent specificity related PRs and so never brought that tweak into the primary section styling PR.

Before the section styling work, only core blocks' style variations could be customized through theme.json or global styles. My thinking is that if a theme or custom block adopts the enhanced block style variations, they might need to tweak their selectors to ensure they play nice with the reduced specificity needed to nest variations.

It might be possible to leave variation styles at the current specificity but I believe long term we'll be best served keeping it to the minimum from the start if possible.

**Comment by aaronrobertshaw** [comment]
I've updated the Image block's rounded style variation but there are a few other core blocks that need further attention. I'll work through those and update again when this is ready for review again.

**Comment by tellthemachines** [comment]
> I've updated the Image block's rounded style variation but there are a few other core blocks that need further attention. I'll work through those and update again when this is ready for review again.

Yeah I'm afraid there hasn't been a concerted effort to lower specificity of all block library styles that might conflict with global ones; we've been doing it piecemeal as we find conflicts. We should probably lower specificity on all styles related to margin, padding, color and typography on all blocks (except ones like Gallery that have custom layout logic). Are you thinking of doing all that as part of this PR, or would it be better handled separately?

**Comment by aaronrobertshaw** [comment]
> We should probably lower specificity on all styles related to margin, padding, color and typography on all blocks

A dedicated sweep sounds like a good option to avoid discovering conflicts in bit by bit kind of way. 

> Are you thinking of doing all that as part of this PR, or would it be better handled separately?

I'm a little torn on this. 

The best situation for the extended block style variations is the reduced specificity for their styles. This makes them dependent on this PR, so it would be great not to block it. On the other hand, we don't want to break anything.

My current thinking/hope was that we could only update styles that impact block style variations as part of this PR's scope. That way we limit any breakage there but somewhat keep the block style variations stuff moving.

In other words, not every margin, padding, color, and typography, style on core blocks will be in the scope of this PR. Only those that disrupt a block style variation for now. What do you think?

**Comment by tellthemachines** [comment]
> My current thinking/hope was that we could only update styles that impact block style variations as part of this PR's scope. That way we limit any breakage there but somewhat keep the block style variations stuff moving.

That sounds good! We can do the rest later. Agree this PR shouldn't be blocked by that work.

**Comment by aaronrobertshaw** [comment]
### Quick Update

While working through a few of the block style variation default styles enqueued by core blocks, some edge cases have arisen that may postpone being able to land this reduction in block style variation specificity.

#### Button Block

The [button block's stylesheet](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/button/style.scss?plain=1#L112) adds a default style for the Outline style variation, making the background transparent.

If this selector's specificity is reduced to zero that default transparent background is lost to any global styles for the Button block or button element as these block library styles are loaded before the global styles. 

To fix this, we'd need a means of moving the default background style for the outline style variation to after the main global styles and before the block style variation styles. Some options here might be to introduce a new type of stylesheet for blocks, similar to `theme.scss` e.g. `variation-defaults.scss`. Another possibility could be to extend the block.json `styles` property to contain style data and implement the feature so these styles are generated and enqueued for both classic and block themes.

#### Social Links

The social links block uses the `Button` component in the editor. This component sets padding styles that would override the default styles for some block style variations if they are reduced to zero specificity.

One option is that we could add an "unstyled" variant to the button component that doesn't include any styling at all. 

#### Table and Separator blocks

I'm still working through the default variation styles for the Table and Separator blocks so there may still be some additional edge cases to add to this list.

**Comment by aaronrobertshaw** [comment]
I've worked through the block library's styles that touch on block style variations. It needs some further testing before it will be ready for a final review though. 

There have also been some inline comments added to stylesheets where the reduction of specificity doesn't make sense. I'm happy to revert these if people don't think they add value. At the least, while reviewing they should flag that the selectors in question have been considered.

The blocks in the library containing block styles include:

- Button
- Image
- Pullquote
- Separator
- Site Logo
- Social Link/s
- Table
- Tag Cloud

The Social Links styles have been updated to be compatible with the zero specificity of global block style variations. This includes fixing a regression introduced by https://github.com/WordPress/gutenberg/pull/60106.

**Important: Reducing the Social Links block's selectors to zero allows bleed from the List block's styles due to it's element based selector `ul,ol`. There is currently an open PR to add a classname to the List block (https://github.com/WordPress/gutenberg/pull/56469) which would resolve the issue here.**

**Comment by tellthemachines** [comment]
> Important: Reducing the Social Links block's selectors to zero allows bleed from the List block's styles due to it's element based selector `ul,ol`

This should now be fixed with #56469 merged!

File: packages/block-library/src/separator/editor.scss [hunk 1]
```diff
@@ -2,11 +2,4 @@
 	// Prevent margin collapsing so the area to select the separator is bigger.
 	padding-top: 0.1px;
 	padding-bottom: 0.1px;
-
-	// This is also set in style.scss, but needs a higher specificity in editor
-	// due to the way that color block supports adds additional background color styles.
-	&.wp-block-separator.is-style-dots {
```
Comment (MaggieCabrera) [inline packages/block-library/src/separator/editor.scss:8]: this is already in the frontend css so this is just clean up 👍

File: phpunit/class-wp-theme-json-test.php [hunk 2]
```diff
@@ -4146,9 +4146,9 @@ public function data_get_styles_for_block_with_style_variations() {
 		$plain = array(
 			'metadata' => array(
 				'path'     => array( 'styles', 'blocks', 'core/quote', 'variations', 'plain' ),
-				'selector' => '.is-style-plain.is-style-plain.wp-block-quote',
+				'selector' => '.is-style-plain.wp-block-quote',
 			),
-			'styles'   => '.is-style-plain.is-style-plain.wp-block-quote{background-color: hotpink;}',
```
Comment (MaggieCabrera) [inline phpunit/class-wp-theme-json-test.php:4151]: Love ❤️

**Comment by aaronrobertshaw** [comment]
There appears to be one remaining issue with this PR. The Gutenberg [reset CSS](https://github.com/WordPress/gutenberg/blob/7d6321997d436de35fba5125c618be43a4016436/packages/block-library/src/reset.scss#L61-L62) generates `0-0-1` specificity styles that currently override the zero specificity styles for block types.

The selectors for these resets are only matched when the editor is iframed e.g. for classic themes or with old block versions.

**Comment by aaronrobertshaw** [comment]
I've put together a GitHub discussion to help settle on [CSS Specificity for WP 6.6](https://github.com/WordPress/gutenberg/discussions/61810).

**Comment by aaronrobertshaw** [comment]
After some [discussion](https://github.com/WordPress/gutenberg/discussions/61810), and taking into account the limited time before the 6.6 beta, the option to use a `:root` prefix seems the safest bet and appears to have majority support.

I'll update this PR accordingly.

**Comment by aaronrobertshaw** [comment]
The changes from this PR are being rolled into https://github.com/WordPress/gutenberg/pull/61638. Closing in favour of that PR.
